crates/dekaf/src/session.rs

```diff
@@ -477,29 +548,347 @@ impl Session {
+            .assigned_partitions
+            .into_iter()
+            .map(|(name, item)| (self.decrypt_topic_name(name.to_owned().into()).into(), item))
+            .collect();
+
+        let mut new_protocol_assignment = BytesMut::new();
+
+        new_protocol_assignment.put_i16(consumer_protocol_assignment_version);
+        consumer_protocol_assignment_msg.encode(
+            &mut new_protocol_assignment,
+            consumer_protocol_assignment_version,
+        )?;
+        mutable_resp.assignment = new_protocol_assignment.into();
+
+        Ok(mutable_resp)
+    }
+
+    pub async fn delete_group(
+        &mut self,
+        req: messages::DeleteGroupsRequest,
+        header: RequestHeader,
+    ) -> anyhow::Result<messages::DeleteGroupsResponse> {
+        return self.app.kafka_client.send_request(req, Some(header)).await;
+    }
+
+    pub async fn heartbeat(
+        &mut self,
+        req: messages::HeartbeatRequest,
+        header: RequestHeader,
+    ) -> anyhow::Result<messages::HeartbeatResponse> {
+        let client = self
+            .app
+            .kafka_client
+            .connect_to_group_coordinator(req.group_id.as_str())
+            .await?;
+        return client.send_request(req, Some(header)).await;
+    }
+
+    pub async fn offset_commit(
+        &mut self,
+        req: messages::OffsetCommitRequest,
+        header: RequestHeader,
+    ) -> anyhow::Result<messages::OffsetCommitResponse> {
+        let mut mutated_req = req.clone();
+        for topic in &mut mutated_req.topics {
+            topic.name = self.encrypt_topic_name(topic.name.clone())
+        }
+
+        let mut resp = self
+            .app
+            .kafka_client
+            .connect_to_group_coordinator(req.group_id.as_str())
+            .await?
+            .send_request(mutated_req, Some(header))
+            .await?;
+
+        for topic in resp.topics.iter_mut() {
+            topic.name = self.decrypt_topic_name(topic.name.to_owned());
+        }
+
+        Ok(resp)
+    }
+
+    #[instrument(skip_all)]
+    pub async fn offset_fetch(
+        &mut self,
+        req: messages::OffsetFetchRequest,
+        header: RequestHeader,
+    ) -> anyhow::Result<messages::OffsetFetchResponse> {
+        let mut mutated_req = req.clone();
+        if let Some(ref mut topics) = mutated_req.topics {
+            for topic in topics {
+                topic.name = self.encrypt_topic_name(topic.name.clone())
             }
         }
 
-        Ok(DescribeConfigsResponse::builder()
-            .results(results)
-            .build()
-            .unwrap())
+        if let Some(ref topics) = mutated_req.topics {
+            self.app
+                .kafka_client
+                .ensure_topics(topics.iter().map(|t| t.name.to_owned()).collect())
+                .await?;
+        }
+        let mut resp = self
+            .app
+            .kafka_client
+            .connect_to_group_coordinator(req.group_id.as_str())
```

Comment:
Should we cache the group ID on the session, along with the routed connection / pool of the current coordinator?

Seems like we could do this once if the session has a long life-cycle for heartbeats and whatnot. The group ID shouldn't ever change for a single session.

Reply:
tl;dr we used to cache the group->coordinator map, and I witnessed the coordinator of a group change which broke that session until I restarted Dekaf. Granted, this happened when I was updating the MSK cluster config so probably not common, but it does happen. These requests are cheap and fast, and I updated to cache the broker uri -> pool map instead